File: runelite-client/src/main/java/net/runelite/client/plugins/alfred/api/rs/inventory/RSInventoryHelper.java
package net.runelite.client.plugins.alfred.api.rs.inventory;

import net.runelite.api.widgets.Widget;
import net.runelite.api.widgets.WidgetInfo;
import net.runelite.client.plugins.alfred.Alfred;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class RSInventoryHelper {

    private final int INVENTORY_CONTAINER_ID = 9764864;
//    private final int UNKNOWN_WIDGET_ID = 786433;

    public boolean isOpen() {
        return Alfred.Companion.getApi().getTabs().getCurrentTab() == WidgetInfo.FIXED_VIEWPORT_INVENTORY_TAB;
//        Widget widget = Alfred.Companion.getApi().getWidgets().getWidget(UNKNOWN_WIDGET_ID);
//        Alfred.Companion.getApi().getTabs().getCurrentTab() == WidgetInfo.FIXED_VIEWPORT_INVENTORY_TAB;
//        if (widget == null) {
//            return false;
//        }
//
//        return !widget.isHidden() && !widget.isSelfHidden();
    }

    public void open() {
        if (!isOpen()) {
            Alfred.Companion.getApi().getTabs().clickInventoryTab();
            Alfred.Companion.sleepUntil(this::isOpen, 100, 2000);
        }
    }

    public int count() {
        return Alfred.Companion.getClientThread().invokeOnClientThread(() -> {
            open();
            Widget inventoryWidget = Alfred.Companion.getApi().getWidgets().getWidget(INVENTORY_CONTAINER_ID);
            int itemCount = 0;

            for (Widget item : inventoryWidget.getDynamicChildren()) {
                if (item == null) {
                    continue;
                }

                if (item.isHidden() || item.isSelfHidden()) {
                    continue;
                }

                if (!item.getName().isEmpty()) {
                    itemCount++;
                }
            }
            return itemCount;
        });
    }

    public boolean isFull() {
        return count() == 28;
    }

    public boolean isEmpty() {
        return count() == 0;
    }

    private boolean internalClickSlot(int slot, boolean leftClick) {
        int correctSlot = slot - 1;

        open();

        Widget slotWidget = Alfred.Companion.getApi().getWidgets().getChildWidget(INVENTORY_CONTAINER_ID, correctSlot);
        if (slotWidget == null) {
            return false;
        }

        if (leftClick) {
            Alfred.Companion.getMouse().leftClick(slotWidget.getBounds());
        } else {
            Alfred.Companion.getMouse().rightClick(slotWidget.getBounds());
        }

        return true;
    }


    public boolean leftClickSlot(int slot) {
        return internalClickSlot(slot, true);
    }

    public boolean rightClickSlot(int slot) {
        return internalClickSlot(slot, false);
    }

    public RSInventoryItem getItemFromSlot(int slot) {
        int correctSlot = slot - 1;

        Widget item = Alfred.Companion.getApi().getWidgets().getChildWidget(INVENTORY_CONTAINER_ID, correctSlot);

        if (item == null) {
            return null;
        }

        return new RSInventoryItem(item);
    }

    private List<RSInventoryItem> internalGetItems() {
        return Alfred.Companion.getClientThread().invokeOnClientThread(() -> {
            open();

            List<RSInventoryItem> rsInventoryItemList = new ArrayList<>();
            Widget inventoryWidget = Alfred.Companion.getApi().getWidgets().getWidget(INVENTORY_CONTAINER_ID);

            for (Widget item : inventoryWidget.getDynamicChildren()) {
                rsInventoryItemList.add(new RSInventoryItem(item));
            }

            return rsInventoryItemList;
        });
    }

    public List<RSInventoryItem> getItems() {
        return internalGetItems();
    }

    public List<RSInventoryItem> getItems(int itemId) {
        return internalGetItems().stream().filter(item -> item.getId() == itemId).collect(Collectors.toList());
    }

    public boolean containsItem(int itemId) {
        return internalGetItems().stream().anyMatch(item -> item.getId() == itemId);
    }

    private boolean internalDrop(RSInventoryItem rsInventoryItem) {
        int currentCount = count();
        boolean success = rsInventoryItem.interact("drop");
        if (!success) {
            return false;
        }
        return Alfred.Companion.sleepUntil(() -> count() == currentCount - 1, 100, 5000);
    }

    private boolean internalDropAll(List<RSInventoryItem> rsInventoryItems) {
        for (RSInventoryItem rsInventoryItem : rsInventoryItems) {
            internalDrop(rsInventoryItem);
        }
        return true;
    }

    private boolean internalDropSlot(int slot) {
        if (slot < 1 || slot > 28) {
            return false;
        }

        if (isEmpty()) {
            return true;
        }

        RSInventoryItem rsInventoryItem = getItemFromSlot(slot);
        internalDrop(rsInventoryItem);
        return true;
    }

    public boolean drop(int itemId) {
        RSInventoryItem rsInventoryItem = Alfred.Companion.getApi().getInventory().getItems(itemId).stream().findFirst().orElse(null);
        if (rsInventoryItem == null) {
            return false;
        }
        return internalDrop(rsInventoryItem);
    }

    public boolean dropFirst(int itemId) {
        RSInventoryItem rsInventoryItem = getItems(itemId).stream().findFirst().orElse(null);
        if (rsInventoryItem == null) return false;
        return internalDrop(rsInventoryItem);
    }

    public boolean dropAll(int itemId) {
        return internalDropAll(getItems(itemId));
    }

    public boolean dropSlot(int slot) {
        return internalDropSlot(slot);
    }

    public boolean dropSlots(int... slots) {
        for (int slot : slots) {
            internalDropSlot(slot);
        }
        return true;
    }

    public boolean dropAllBetween(int start, int end) {
        for (int i = start; i <= end; i++) {
            internalDropSlot(i);
        }
        return true;
    }

    private boolean internalInteract(RSInventoryItem rsInventoryItem, String action) {
        if (rsInventoryItem == null) {
            return false;
        }

        if (!rsInventoryItem.getInventoryActions().contains(action)) {
            return false;
        }

        return rsInventoryItem.interact(action);
    }


    public boolean interactFirst(int itemId, String action) {
        RSInventoryItem rsInventoryItem = getItems(itemId).stream().findFirst().orElse(null);
        return internalInteract(rsInventoryItem, action);
    }

    private boolean internalInteractAll(RSInventoryItem rsInventoryItem, String action) {
        List<RSInventoryItem> rsInventoryItems = getItems(rsInventoryItem.getId());
        for (RSInventoryItem item : rsInventoryItems) {
            internalInteract(item, action);
            Alfred.Companion.sleep(75, 150);
        }
        return true;
    }

    public boolean interactAll(int itemId, String action) {
        RSInventoryItem rsInventoryItem = getItems(itemId).stream().findFirst().orElse(null);
        if (rsInventoryItem == null) {
            return false;
        }
        return internalInteractAll(rsInventoryItem, action);
    }
}

File: runelite-client/src/main/java/net/runelite/client/plugins/alfred/api/rs/inventory/RSInventoryHelper.kt
package net.runelite.client.plugins.alfred.api.rs.inventory

import net.runelite.api.widgets.Widget
import net.runelite.api.widgets.WidgetInfo
import net.runelite.client.plugins.alfred.Alfred

class RSInventoryHelper {
    private val INVENTORY_CONTAINER_ID = 9764864
    val isOpen: Boolean
        //    private final int UNKNOWN_WIDGET_ID = 786433;
        get() = Alfred.api.tabs.currentTab == WidgetInfo.FIXED_VIEWPORT_INVENTORY_TAB

    //        Widget widget = Alfred.Companion.getApi().getWidgets().getWidget(UNKNOWN_WIDGET_ID);
//        Alfred.Companion.getApi().getTabs().getCurrentTab() == WidgetInfo.FIXED_VIEWPORT_INVENTORY_TAB;
//        if (widget == null) {
//            return false;
//        }
//
//        return !widget.isHidden() && !widget.isSelfHidden();
    fun open() {
        if (!isOpen) {
            Alfred.api.tabs.clickInventoryTab()
            Alfred.sleepUntil({ isOpen }, 100, 2000)
        }
    }

    fun count(): Int {
        return Alfred.clientThread.invokeOnClientThread {
            open()
            val inventoryWidget = Alfred.api.widgets.getWidget(INVENTORY_CONTAINER_ID)
            return@invokeOnClientThread inventoryWidget!!.dynamicChildren.filterNotNull()
                .filter { widget: Widget -> !widget.isHidden && !widget.isSelfHidden }
                .count { widget: Widget -> widget.name.isNotEmpty() && widget.name.isNotBlank() }
        }
    }

    val isFull: Boolean
        get() = count() == 28
    val isEmpty: Boolean
        get() = count() == 0

    private fun internalClickSlot(slot: Int, leftClick: Boolean): Boolean {
        val correctSlot = slot - 1
        open()
        val slotWidget = Alfred.api.widgets.getChildWidget(INVENTORY_CONTAINER_ID, correctSlot) ?: return false
        if (leftClick) {
            Alfred.mouse.leftClick(slotWidget.getBounds())
        } else {
            Alfred.mouse.rightClick(slotWidget.getBounds())
        }
        return true
    }

    fun leftClickSlot(slot: Int): Boolean {
        return internalClickSlot(slot, true)
    }

    fun rightClickSlot(slot: Int): Boolean {
        return internalClickSlot(slot, false)
    }

    fun getItemFromSlot(slot: Int): RSInventoryItem? {
        val correctSlot = slot - 1
        val item = Alfred.api.widgets.getChildWidget(INVENTORY_CONTAINER_ID, correctSlot) ?: return null
        return RSInventoryItem(item)
    }

    private fun internalGetItems(): List<RSInventoryItem> {
        return Alfred.clientThread.invokeOnClientThread {
            open()
            val inventoryWidget = Alfred.api.widgets.getWidget(INVENTORY_CONTAINER_ID)
            return@invokeOnClientThread inventoryWidget!!.dynamicChildren.filterNotNull().map { widget: Widget -> RSInventoryItem(widget) }
        }
    }

    val items: List<RSInventoryItem?>
        get() = internalGetItems()

    fun getItems(itemID: Int): List<RSInventoryItem> {
        return internalGetItems().filter { rsInventoryItem: RSInventoryItem -> rsInventoryItem.id == itemID }
    }

    fun containsItem(itemID: Int): Boolean {
        return internalGetItems().any { rsInventoryItem: RSInventoryItem -> rsInventoryItem.id == itemID }
    }

    private fun internalDrop(rsInventoryItem: RSInventoryItem): Boolean {
        val currentCount = count()
        val success = rsInventoryItem.interact("drop")

        return if (!success) {
            false
        } else Alfred.sleepUntil({ count() == currentCount - 1 }, 100, 5000)
    }

    private fun internalDropAll(rsInventoryItems: List<RSInventoryItem>): Boolean {
        rsInventoryItems.forEach { rsInventoryItem ->
            internalDrop(rsInventoryItem)
        }
        return true
    }

    private fun internalDropSlot(slot: Int): Boolean {
        if (slot < 1 || slot > 28) {
            return false
        }
        if (isEmpty) {
            return true
        }
        val rsInventoryItem = getItemFromSlot(slot) ?: return false
        return internalDrop(rsInventoryItem)
    }

    fun drop(itemId: Int): Boolean {
        val rsInventoryItem = Alfred.api.inventory.getItems(itemId).firstOrNull() ?: return false
        return internalDrop(rsInventoryItem)
    }

    fun dropFirst(itemId: Int): Boolean {
        val rsInventoryItem = getItems(itemId).firstOrNull() ?: return false
        return internalDrop(rsInventoryItem)
    }

    fun dropAll(itemId: Int): Boolean {
        return internalDropAll(getItems(itemId))
    }

    fun dropSlot(slot: Int): Boolean {
        return internalDropSlot(slot)
    }

    fun dropSlots(vararg slots: Int): Boolean {
        for (slot in slots) {
            internalDropSlot(slot)
        }
        return true
    }

    fun dropAllBetween(start: Int, end: Int): Boolean {
        for (i in start..end) {
            internalDropSlot(i)
        }
        return true
    }

    private fun internalInteract(rsInventoryItem: RSInventoryItem, action: String): Boolean {
        return if (!rsInventoryItem.inventoryActions.contains(action)) {
            false
        } else rsInventoryItem.interact(action)
    }

    fun interactFirst(itemID: Int, action: String): Boolean {
        val rsInventoryItem = getItems(itemID).firstOrNull() ?: return false
        return internalInteract(rsInventoryItem, action)
    }

    private fun internalInteractAll(rsInventoryItem: RSInventoryItem, action: String): Boolean {
        val rsInventoryItems = getItems(rsInventoryItem.id)
        rsInventoryItems.forEach { item ->
            internalInteract(rsInventoryItem, action)
            Alfred.sleep(75, 150)
        }
        return true
    }

    fun interactAll(itemId: Int, action: String): Boolean {
        val rsInventoryItem = getItems(itemId).firstOrNull() ?: return false
        return internalInteractAll(rsInventoryItem, action)
    }
}
